pygfx/objects/_base.py

```diff
@@ -306,63 +294,92 @@ def receive_shadow(self, value):
-        An object can have at most one parent.
-        """
-        return self._parent_ref and self._parent_ref()
+    def parent(self) -> "WorldObject":
+        if self._parent is None:
+            return None
+        else:
+            return self._parent()
+
+    def add(self, *objects, before=None, keep_world_matrix=False) -> "WorldObject":
+        """Add child objects.
+
+        Any number of objects may be added. Any current parent on an object
+        passed in here will be removed, since an object can have at most one
+        parent. If ``before`` argument is given, then the items are inserted
+        before the given element.
+
+        Parameters
+        ----------
+        *objects : WorldObject
+            The world objects to add as children.
+        before : WorldObject
+            If not None, insert the objects before this child object.
+        keep_world_matrix : bool
+            If True, the child will keep it's world transform. It moves in the
+            scene graph but will visually remain in the same place. If False,
+            the child will keep it's parent transform.
 
-    @property
-    def children(self):
-        """The child objects of this wold object (read-only tuple).
-        Use ``.add()`` and ``.remove()`` to change this list.
-        """
-        return tuple(self._children)
-
-    def add(self, *objects, before=None):
-        """Adds object as child of this object. Any number of
-        objects may be added. Any current parent on an object passed
-        in here will be removed, since an object can have at most one
-        parent.
-        If ``before`` argument is given (and present in children), then
-        the items are inserted before the given element.
         """
-        idx = len(self._children)
-        if before:
-            try:
-                idx = self._children.index(before)
-            except ValueError:
-                pass
+
+        obj: WorldObject
         for obj in objects:
-            assert isinstance(obj, WorldObject)
-            # orphan if needed
-            if obj._parent_ref is not None:
-                obj._parent_ref().remove(obj)
-            # attach to scene graph
-            obj._parent_ref = weakref.ref(self)
-            self._children.insert(idx, obj)
-            idx += 1
-            # flag world matrix as dirty
-            obj._matrix_world_dirty = True
+            if obj.parent is not None:
+                obj.parent.remove(obj, keep_world_matrix=keep_world_matrix)
+
+            if before is not None:
+                idx = self.children.index(before)
+            else:
+                idx = len(self.children)
+
+            if keep_world_matrix:
+                transform_matrix = obj.world.matrix
+
+            obj._parent = weakref.ref(self)
+            obj.world.parent = self.world
+            self.children.insert(idx, obj)
+
+            if keep_world_matrix:
+                obj.world.matrix = transform_matrix
```

Comment:
does this also lead to an updated `local` transform? or will they be out of sync?

Reply:
Yes, `obj.local` will also be updated. `obj.world` is a `RecursiveTransform` which is initialized to use `obj.local` as its `self.own` matrix. As a result, any updates (`__set__`) of `obj.world` properties will be propagated to `obj.local` after having been converted to parent space.

E.g., `obj.world.matrix = transform_matrix` will take `transform_matrix` left-multiply it with the inverse of `parent` (this removes everything except the local->parent transform) and then assign the result to `obj.local.matrix`.

Reply:
Great! Ok.